src/lib/tls/tls13/tls_client_impl_13.cpp

```diff
@@ -130,12 +130,37 @@ bool Client_Impl_13::handshake_finished() const
 
 std::optional<Session_with_Handle> Client_Impl_13::find_session_for_resumption()
    {
+   auto sessions = session_manager().find(m_info, callbacks(), policy());
+   if(sessions.empty())
+      { return std::nullopt; }
+
    // TODO: TLS 1.3 allows sending more than one ticket (for resumption) in a
    //       Client Hello. Currently, we do not support that. The Session_Manager
    //       does imply otherwise with its API, though.
-   if(auto sessions = session_manager().find(m_info, callbacks(), policy()); !sessions.empty())
-      return sessions.front();
-   return std::nullopt;
+   auto& session_to_resume = sessions.front();
+
+   // RFC 8446 4.6.1
+   //    Clients MUST only resume if the new SNI value is valid for the
+   //    server certificate presented in the original session and SHOULD only
+   //    resume if the SNI value matches the one used in the original session.
+   //
+   // ... i.e. we do not attempt a resumption if the session's certificate does
+   // not withstand basic validity checks or -- if no server certificate is
+   // available -- at least the session's host name matches the expectations.
+   const auto& cert_chain = session_to_resume.session.peer_certs();
+   if(!cert_chain.empty())
+      {
+      const auto& server_cert = cert_chain.front();
+      if(!server_cert.is_self_signed() &&
```

Comment:
I've gotten this idiom from somewhere else. Frankly I don't remember from where.
Nevertheless, I found it reasonable, as self-signed certs don't come with any security guarantee anyway. But I do agree that redoing this check for a resumption attempt probably isn't breaking anything either.

Reply:
I would say we should always do the check, as the RFC does not mention any exception for self signed certs.

Reply:
I introduced this check because of the RFC paragraph referenced in the code. The exception for self-signed certificates was (presumably) introduced to pass BoGo tests (that use certificates without SNI/hostname information).

I came to the conclusion that we should not hard-code this check at all.

1. **SNI validation against the original session's certificate**
  Actual validation of the certificate is usually done via `Callbacks::tls_verify_cert_chain()`. In this callback applications can override any SNI restrictions in an ordinary handshake. Therefore, we cannot assume that a hard-coded SNI check for certificates in the resumption case would actually succeed (for instance BoGo tests would fail).
  Instead, we might re-call `tls_verify_cert_chain()` for the stored server cert chain before resumption. However, the RFC does not request a full re-validation of certificates.

2. **Equality of new SNI hostname and the one stored in the session**
  This is practically a given. We are querying the `Session_Manager` for the SNI value. Assuming that the manager does not lie to us, we don't need to re-validate that here.
  
Conclusion: Both checks are out of place here and can be implemented by an application (in a custom session manager) if needed.